cpp/src/arrow/compute/kernels/vector_swizzle_test.cc

```diff
@@ -0,0 +1,1065 @@
+}
+
+void DoTestScatterForIndicesTypes(
+    const std::vector<std::shared_ptr<DataType>>& indices_types,
+    const std::shared_ptr<Array>& values, const std::shared_ptr<Array>& indices,
+    int64_t max_index, const std::shared_ptr<Array>& expected) {
+  for (const auto& indices_type : indices_types) {
+    ARROW_SCOPED_TRACE("Indices type: " + indices_type->ToString());
+    ASSERT_OK_AND_ASSIGN(auto casted, Cast(indices, indices_type));
+    ASSERT_TRUE(casted.is_array());
+    auto casted_indices = casted.make_array();
+    {
+      ARROW_SCOPED_TRACE("AAA");
+      DoTestScatterAAA(values, casted_indices, max_index, expected);
+    }
+    {
+      ARROW_SCOPED_TRACE("CAA");
+      DoTestScatterCACWithArrays(values, casted_indices, max_index, expected);
+    }
+    {
+      ARROW_SCOPED_TRACE("ACA");
+      DoTestScatterACCWithArrays(values, casted_indices, max_index, expected);
+    }
+    {
+      ARROW_SCOPED_TRACE("CCA");
+      DoTestScatterCCCWithArrays(values, casted_indices, max_index, expected);
+    }
+  }
+}
+
+void DoTestScatterSignedIndices(const std::shared_ptr<Array>& values,
+                                const std::shared_ptr<Array>& indices, int64_t max_index,
+                                const std::shared_ptr<Array>& expected) {
+  DoTestScatterForIndicesTypes(kSignedIntegerTypes, values, indices, max_index, expected);
+}
+
+void DoTestScatter(const std::shared_ptr<Array>& values,
+                   const std::shared_ptr<Array>& indices, int64_t max_index,
+                   const std::shared_ptr<Array>& expected) {
+  DoTestScatterForIndicesTypes(kIntegerTypes, values, indices, max_index, expected);
+}
+
+}  // namespace
+
+TEST(Scatter, Invalid) {
+  {
+    ARROW_SCOPED_TRACE("Length mismatch");
+    auto values = ArrayFromJSON(int32(), "[0, 1]");
+    auto indices = ArrayFromJSON(int32(), "[0]");
+    ASSERT_RAISES_WITH_MESSAGE(
+        Invalid,
+        "Invalid: Input and indices of scatter must have the same length, got 2 and 1",
+        Scatter(values, indices));
+  }
+  {
+    ARROW_SCOPED_TRACE("Invalid input type");
+    auto values = ArrayFromJSON(int32(), "[0]");
+    auto indices = ArrayFromJSON(utf8(), R"(["a"])");
+    ASSERT_RAISES_WITH_MESSAGE(
+        Invalid, "Invalid: Indices of scatter must be of integer type, got string",
+        Scatter(values, indices));
+  }
+}
+
+TEST(Scatter, DefaultOptions) {
+  {
+    ARROW_SCOPED_TRACE("Default options values");
+    ScatterOptions options;
+    ASSERT_EQ(options.max_index, -1);
+  }
+  {
+    ARROW_SCOPED_TRACE("Default options semantics");
+    auto values = ArrayFromJSON(utf8(), R"(["a"])");
+    for (const auto& indices_type : kIntegerTypes) {
+      ARROW_SCOPED_TRACE("Indices type: " + indices_type->ToString());
+      auto indices = ArrayFromJSON(indices_type, "[0]");
+      ASSERT_OK_AND_ASSIGN(Datum result, Scatter(values, indices));
+      AssertDatumsEqual(values, result);
+    }
+  }
+}
+
+template <typename ArrowType>
```

Comment:
Templated tests are convenient but they also tend to increase compilation times. When doable, it's preferrable to use value parameterization (or simple for loops!). This is the case here thanks to `ArrayFromJSON`: for most tests, you can iterate over the different input/output types.

Reply:
Good point. I've reduced typed tests that are the case by using for loops. However I tend to keep the remaining two typed tests (`TestInversePermutationSmallOutputType` and `TestScatterSmallIndicesTypes`) as are, because: 1) I need to apply `std::numeric_limits` by the type parameter; 2) the number of "small types" are reduced to 2 so the code generation should be acceptable.

Reply:
This sounds reasonable indeed, thanks.